### ISSUE TITLE
CMake selecting wrong MSVC compiler, Windows 10

### ISSUE THREAD

**Issue by mjscosta** [scikit-build/scikit-build#455]
Hello, 

Environment,

conda environment using:

cmake v 3.15.5

Hope someone can shed some light into the following issue.

I'm building a python native extension using MSVC Build Tools 2015 installed as a package for MSVC Build Tools 2017.

To build I follow the steps:

1 - Activate compiler environment for my specific configuration, using 
`call "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Auxiliary\Build\vcvarsall.bat" amd64  -vcvars_ver=14.0
`
2- Execute setup.py
`python setup.py bdist_wheel --build-type Release -G "NMake Makefiles"
`
3- Results in:

```
--------------------------------------------------------------------------------
-- Trying "NMake Makefiles" generator
--------------------------------
---------------------------
----------------------
-----------------
------------
-------
--
Not searching for unused variables given on the command line.
-- The C compiler identification is MSVC 19.16.27034.0
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- The CXX compiler identification is MSVC 19.16.27034.0
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
```
Which leads to later to an error because I need to build with the toolset v 140.
```
fatal error C1047: The object or library file '************************************' was created with an older compiler than other objects; rebuild old objects and libraries
```
3.1 Inspecting the compiler in the command line:
```
>where cl
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64\cl.exe

>cl
Microsoft (R) C/C++ Optimizing Compiler Version 19.00.24234.1 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

>"C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe"
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27034 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.
```
4- For some reason, cmake via scikit-build is not accepting the command line environment for MSVC 2015, instead it picks up the MSVC 2017 compiler.
5- The same project also has a pure C++ build, for which I invoke cmake from the command line, in such case with the exact same generator cmake uses the command line environment and picks up the right C++ compiler. So I assume there's some issue related to scikit-build.

Is there any way to force scikit-build to use the MSVC compiler from the command line ?

Thanks in advance!

**Comment by mjscosta** [comment]
I've managed to find the reason for this different behavior.

1.1 - scikit-build assumes that the compiler v141 for python >= 3.6 - # For Python 3.6 and above: VS2017:
https://github.com/scikit-build/scikit-build/blob/d2f0419c0f3283e33da45dd8f74dcc6743671d97/skbuild/platform_specifics/windows.py#L78

1.2 - scikit-build does not reuse the process environment when calling cmake instead uses generator.env:
 https://github.com/scikit-build/scikit-build/blob/b2d4bc29b169a5b5f8a5b48de411853c06847630/skbuild/cmaker.py#L227

2- Stating my case, in the "official python page is mentions 14.X for Python 3.5, 3.6, 3.7, 3.8.
https://wiki.python.org/moin/WindowsCompilers
2.1 - The python 3.6 I'm using from conda-forge which is built with the toolchain v140, and as it seems in some scenarios it is incompatible with v141.
https://conda-forge.org/docs/maintainer/knowledge_base.html

At least v140 should be possible to use with python 3.6, as for conda-forge, they maintain full packaging tree for each python version, and C++ and python packages with native extensions need to be compatible with python 3.6 that's compiled with MSVC v140 ...

Any change this could be improved to support MSVC v140 for other python versions ?

**Comment by thewtex** [comment]
@mjscosta well done on the investigation :+1: 

> Any change this could be improved to support MSVC v140 for other python versions ?

Could you please submit a patch? We want to be compatible with conda-forge.

**Comment by mjscosta** [comment]
Hi @thewtex , 
Sure, I'm looking into it, not sure whats the best approach, maybe you can help-me clarify some of my questions.

1- scikit-build - has a lot of code focused on testing and finding the VS Compiler, instead of relying on just the available environment variables (the way cmake works), and delegating to cmake the finding of the compiler and also the testing of the compiler, hence delegating the responsibility to the caller to setup the environment.

1.1 Instead it is making assumptions about the compilers to use, which on some specific scenarios might be different for some reason/use case ...

1.2 In the case of "Ninja" Generators and "NMake" generators, there's no "-T" option to pass the tool-set (v140/v141).

2- Found some other issues, namely missing support for upcoming MSVC 2019, that has now replaced 2017, which will also break on the vswhere query. (adding a list of components bellow that need support.)

Build Tools 2019
components -
Microsoft.VisualStudio.Component.VC.Tools.x86.x64 - v142 - tool-set
Microsoft.VisualStudio.Component.VC.v141.x86.x64 - v141 - tool-set
Microsoft.VisualStudio.Component.VC.140 - v140 - tool-set

Build Tools 2017
Microsoft.VisualStudio.Component.VC.Tools.x86.x64 - v141 tool-set
Microsoft.VisualStudio.Component.VC.140 - v140 tool-set

... Sorry, missing the questions.

3 - The important question for me is, why is scikit-build setting the compiler environment ?

Possible approaches:

4.1 - Keep compatibility with previous behavior and add an option to just ignore all the generator tests and selection of cached generators and just inherit the process environment on the call to cmake by using `env=None` in the `subprocess `call.
 This one would delegate to the caller the responsibility of setting the proper compiler environment before calling setup.py (same as cmake approach).
 
4.2 - Extend the functionality to:
4.2.1 - Support more versions of compilers and toolsets per python version
4.2.2 - Add a means to hint for a preferred generator (MSVC version and tool-set), right now if there's more than one available, the first one is selected which might not be the desired one.

**Comment by thewtex** [comment]
Some background: in general the choice of toolchain and toolchain version is selected so Python binary extensions are compatible with the the binary Python distributed at python.org. In general, this means using the same C runtime that the toolchain uses.

When scikit-build was created, every version of CPython on Windows was compiled with a different version of Visual Studio, which had an incompatible C runtime. More recently, with Visual Studio 2015 and the 14.X toolsets that Python 3.5 and later use, Microsoft has used the Universal C Runtime, *ucrt*. In practice, this seems to achieve its intended goal of being compatible across toolsets.

So, in general, we want scikit-build by default to pick the same or a compatible toolchain for the C Python it is building against. We can pick good defaults based on the Python version. However, we want the ability to override those defaults for cases like conda-forge, which may have used a different toolchain, or when using a compatible toolchain if the version python.org did not use is not available, e.g. a different version of the recent ucrt-compatible toolchains on Windows. This is explained here:

  https://scikit-build.readthedocs.io/en/latest/generators.html

Essentially, what we want is a variant of your 4.2.

And, scikit-build facilitates passing arguments to CMake via the call to setup.py, so compilers, toolchain, generators, toolsets can / should be able to be specified via `CMAKE_C_COMPILER`, `CMAKE_TOOLCHAIN_FILE`, `CMAKE_CXX_COMPILER`, `-G`, `-T`, etc.

  https://scikit-build.readthedocs.io/en/latest/usage.html#cmake-configure-options

That's the idea, at least :-).

**Comment by mjscosta** [comment]
Thanks a lot for your clarification, I had indeed the my questions answered.

The issue seems to rise in a very specific scenario, as described here: https://bugs.python.org/issue31340#msg303884

And it rises only when linking static build libraries previously build with a different toolset, which is my exact scenario ...

I've spent some time investigating the code and drafting a solution for 4.2, need to do a little bit more thinking to have something that fits properly and does not break previous behavior.

**Comment by T4mmi** [comment]
Hi there,
Not sure about everything said previously but just adding a personnal experience 
>2- Found some other issues, namely missing support for upcoming MSVC 2019, that has now replaced 2017, which will also break on the vswhere query.

Indeed, I have MS Build Tools 2019 installed which is detected by `cmake` and the `setupools` approach using:
```
class CMakeExtension(Extension):
    def __init__(self, name, sourcedir=""):
        Extension.__init__(self, name, sources=[])
        self.sourcedir = os.path.abspath(sourcedir)

class CMakeBuild(build_ext):
    def run(self):
        [...]

    def build_extension(self, ext):
        [...]
```

but I have to manualy force the Visual Studio 2019 generators to use scikit-build (`python setup.py build -G "Visual Studio 16 2019"`) otherwise it fails:
```
scikit-build could not get a working generator for your system. Aborting build.
```

**Comment by jcfr** [comment]
Update : We are currently working on finalizing https://github.com/scikit-build/scikit-build/pull/526 adding support for VS2019 along with ensuring that toolset and architecture argument are passed to the project being configured.

**Comment by ZzEeKkAa** [comment]
I'm still facing issues with setting up specific toolset while using DPCPP compiler. The problem is that it is compatible with specific version of MSVC toolset. That means that `vcvars64.bat -vcvars_ver=14.3` triggering latest toolset (for the moment of writing, it is 14.37) and it results in failed build, because `dpcpp 2023.2` is compatible with toolsets up to `14.36`. That causes compile time issues on any system that contains `14.37` install. Calling `vcvars64.bat -vcvars_ver=14.36` does not work because `windows.py` overwrites environment variables and they point to the output of `vcvars64.bat -vcvars_ver=14.3` which is incompatible `14.37`.
In other words we need to provide some arguments to specify exact toolset we want to use.
P.S.: we are using **MSVC Ninja** generator.

**Comment by ZzEeKkAa** [comment]
After looking through the code, I guess it will be nice to filter out MSVC Ninja generators base of MSVC environment variable, so it does the check ahead of calling CMake, not during CMake check. It will result in faster Pipelines.

**Comment by ZzEeKkAa** [comment]
For anybody facing similar issues with setting up specific toolset I ended up with monkey patching setup.py.

**Workaround**:

1. Call `vcvars64.bat -vcvars_ver=<vcvars_ver>` with the version you need.
2. In your setup.py monkey patch the code that calls `vcvars64.bat`.
3. Run `setup.py` or `pip install` 


setup.py monkey patch:
```python
from skbuild.platform_specifics import windows

# Monkey patch msvc compiler environment, so scikit-build does not overwrite
# it. Make sure to set desired environment using:
# > vcvars64.bat -vcvars_ver=<vcvars_ver>
windows._get_msvc_compiler_env = lambda version, toolset: windows.CachedEnv()
```

**Comment by timorkal** [comment]
Any progress on this matter? I am trying to build in Visual Studio 2022 with MSVC 2019 Build Tools, and skbuild still tries to automatically select the new MSVC version.
Although the proposed Workaround did work, it looks like not the supposed way to do it.

**Comment by ZzEeKkAa** [comment]
@timorkal I guess the meta right now is to use https://github.com/scikit-build/scikit-build-core , so no development here, unfortunately.

**Comment by henryiii** [comment]
Eventually we'll swap out the backend for scikit-build-core, so it would be interesting to see if that one works. It does have a much simpler discovery system that tends to be more reliable.

**Comment by henryiii** [comment]
scikit-build (classic) after #1185 is now a thin wrapper around scikit-build-core's setuptools plugin. Closing all open issues; if this still applies, please reopen or open an issue on scikit-build-core. Thanks!
